src/backend/distributed/commands/alter_table.c

```diff
@@ -803,11 +803,24 @@ ConvertTable(TableConversionState *con)
 	char *attachPartitionCommand = NULL;
 	foreach_ptr(attachPartitionCommand, attachPartitionCommands)
 	{
+		/*
+		 * when there are many partitions, each call to ProcessUtilityParseTree
+		 * accumulates used memory. Free context after each call.
+		 */
+		MemoryContext citusPerPartitionContext =
+			AllocSetContextCreate(CurrentMemoryContext,
+								  "citus_per_partition_context",
+								  ALLOCSET_DEFAULT_SIZES);
```

Comment:
would be slightly nicer to create the memory context once and reset it in every iteration.